src/mlpack/methods/ann/layer/adaptive_max_pooling_impl.hpp

```diff
@@ -0,0 +1,156 @@
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+AdaptiveMaxPoolingType<MatType>::AdaptiveMaxPoolingType(
+    AdaptiveMaxPoolingType&& other) :
+    Layer<MatType>(std::move(other)),
+    poolingLayer(std::move(other.poolingLayer)),
+    outputWidth(std::move(other.outputWidth)),
+    outputHeight(std::move(other.outputHeight))
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+AdaptiveMaxPoolingType<MatType>&
+AdaptiveMaxPoolingType<MatType>::operator=(const AdaptiveMaxPoolingType& other)
+{
+  if (&other != this)
+  {
+    Layer<MatType>::operator=(other);
+    poolingLayer = other.poolingLayer;
+    outputWidth = other.outputWidth;
+    outputHeight = other.outputHeight;
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+AdaptiveMaxPoolingType<MatType>&
+AdaptiveMaxPoolingType<MatType>::operator=(AdaptiveMaxPoolingType&& other)
+{
+  if (&other != this)
+  {
+    Layer<MatType>::operator=(std::move(other));
+    poolingLayer = std::move(other.poolingLayer);
+    outputWidth = std::move(other.outputWidth);
+    outputHeight = std::move(other.outputHeight);
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+void AdaptiveMaxPoolingType<MatType>::Forward(
+    const MatType& input, MatType& output)
+{
+  poolingLayer.Training() = this->training;
+  poolingLayer.Forward(input, output);
+}
+
+template<typename MatType>
+void AdaptiveMaxPoolingType<MatType>::Backward(
+  const MatType& input,
+  const MatType& gy,
+  MatType& g)
+{
+  poolingLayer.Backward(input, gy, g);
+}
+
+template<typename MatType>
+void AdaptiveMaxPoolingType<MatType>::ComputeOutputDimensions()
+{
+  // The AdaptiveMaxPooling layer only affects the first two dimensions.
+  this->outputDimensions = this->inputDimensions;
+
+  this->outputDimensions[0] = outputWidth;
+  this->outputDimensions[1] = outputHeight;
+
+  if (outputWidth > this->inputDimensions[0] ||
+        outputHeight > this->inputDimensions[1])
+  {
+    Log::Fatal << "Given output shape (" << outputWidth << ", "
+      << outputHeight << ") is not possible for given input shape ("
+      << this->inputDimensions[0] << ", " << this->inputDimensions[1]
+      << ")." << std::endl;
+  }
+  poolingLayer.InputDimensions() = this->inputDimensions;
+  poolingLayer.StrideWidth() = std::floor(this->inputDimensions[0] /
+      outputWidth);
+  poolingLayer.StrideHeight() = std::floor(this->inputDimensions[1] /
+      outputHeight);
+
+  poolingLayer.KernelWidth() = this->inputDimensions[0] -
+      (outputWidth - 1) * poolingLayer.StrideWidth();
+  poolingLayer.KernelHeight() = this->inputDimensions[1] -
+      (outputHeight - 1) * poolingLayer.StrideHeight();
+
+  poolingLayer.ComputeOutputDimensions();
+
+  // Higher dimensions are not modified.
+}
+
+template<typename MatType>
+template<typename Archive>
+void AdaptiveMaxPoolingType<MatType>::serialize(
+    Archive& ar,
+    const uint32_t /* version */)
+{
+  ar(cereal::base_class<Layer<MatType>>(this));
+
```

Comment:
nit: Remove the extra empty line from here

Reply:
I think there is an extra line in every layer.
So I think we should keep it.